query/src/main/scala/filodb/query/exec/rangefn/RateFunctions.scala

```diff
@@ -49,6 +49,71 @@ object RateFunctions {
       Some(result)
     }
   }
+
+  def derivFunction(window: Window): Option[Double] = {
+    if (window.size < 2) {
+      None  // cannot calculate result without 2 samples
+    } else {
+      // We pass in an arbitrary timestamp that is near the values in use
+      // to avoid floating point accuracy issues
+      Some(linearRegression(window, window(0).timestamp)._1)
+    }
+  }
+
+  /**
+    * Logic is kept consistent with Prometheus' linearRegression function in order to get consistent results.
+    * We can look at optimizations (if any) later.
+    */
+  def linearRegression(window: Window, interceptTime: Long): Tuple2[Double, Double] = {
```

Comment:
Tuple[Double, Double] will generate a lot of garbage. Please see if you can avoid this

Reply:
What is the preferred choice here?

Reply:
@vishramachandran "intercept" is needed for `predict_linear` function, but for now making it double. We can add it back when we implement it.